### ISSUE TITLE
Refactor the current structure of hBayesDM with subfunctions

### ISSUE THREAD

**Issue by JaeyeongYang** [CCS-Lab/hBayesDM#43]
Since the current version of hBayesDM has so many common codes between different functions, it seems inefficient to contains redundant codes and hard to find out which parts we should change for new models. Thus, we are going to make some subfunctions to combine the common codes on the current functions into several core subfunctions.

**Comment by dlemfh96** [comment]
```r
model_base <- function(
  model_name,
  model_type,
  example_filename,
  data_names,
  parameter_names,
  regressor_available,
  regressor_names,
  preprocess_function
) { ... }
```
- `model_name` : Name of the model, e.g. `"gng_m1"`.
- `model_type` : One of the following three: `""` OR `"single"` OR `"multipleB"`.
- `example_filename` : Filename of the example data to be provided with this model, e.g. `"gng_exampleData.txt"`.
- `data_names` : Names of the necessary column labels for the data, e.g. `c("subjID", "cue", "keyPressed", "outcome")`. Must be the entirety of necessary data column labels that will be used at some point in the code. I.e. must always include `"subjID"`, and should include `"block"` in the case of 'multipleB' type models.
- `parameter_names` : Names of the parameters of this model, e.g. `c("xi", "ep", "rho")`.
- `regressor_available` : `TRUE` if model-based regressors are available for this model.
- `regressor_names` : Names of the model-based regressors for this model, e.g. `c("Qgo", "Qnogo", "Wgo", "Wnogo")`. Will not be used if `regressor_available == FALSE`.
- `preprocess_function` : The model-specific function to preprocess the raw data to pass to Stan. Takes a data.frame object `raw_data` as argument; and returns a list object `stan_data`.

**Comment by JaeyeongYang** [comment]
- Suggestions on argument names:
    - `data_names` -> `data_columns`
    - `parameter_names` -> `parameters`
    - `regressor_names` -> `regressors`
- Maybe we can omit `regressor_available` by checking the length of `regressors`.

**Comment by dlemfh96** [comment]
### base.R
```r
model_base <- function(
    task_name,
    model_name,
    model_type,
    data_columns,
    parameters,
    regressors,
    preprocess_function
) { ... }
```

### Usage ex.
```r
gng_m1 <- model_base(
    task_name           = "gng",
    model_name          = "m1",
    model_type          = "",
    data_columns        = c("subjID", "cue", "keyPressed", "outcome"),
    parameters          = c("xi", "ep", "rho"),
    regressors          = c("Qgo", "Qnogo", "Wgo", "Wnogo"),
    preprocess_function = function(raw_data) {

            ...

    }
)
```

### Description
- `task_name` : Name of the task, e.g. `"gng"`.
- `model_name` : Name of the model with regards to the task, e.g. `"m1"`.
- `model_type` : One of the following three: `""` OR `"single"` OR `"multipleB"`.
- `data_columns` : Names of the necessary columns for the data, e.g. `c("subjID", "cue", "keyPressed", "outcome")`. Must be the entirety of necessary data columns that will be used at some point in the code. I.e. must always include `"subjID"`, and should include `"block"` in the case of 'multipleB' type models.
- `parameters` : Names of the parameters of this model, e.g. `c("xi", "ep", "rho")`.
- `regressors` : Names of the model-based regressors for this model, e.g. `c("Qgo", "Qnogo", "Wgo", "Wnogo")`. OR `NA` if model_based regressors are not available for this model.
- `preprocess_function` : The model-specific function to preprocess the raw data to pass to Stan. Takes a data.frame object `raw_data` as argument, and returns a list object `stan_data`.

**Comment by JaeyeongYang** [comment]
Good. How about using `""` or `NULL` as a default value for `model_type` argument?

**Comment by dlemfh96** [comment]
### Current progress report
- [x] Delete lines related to measuring `timeTook` because no code in the whole package is currently making use of `timeTook` after it is measured.
- [x] Add procedure to check if user-given data has all the correctly labeled necessary data columns, and `stop()` the process if it does not.
- [x] Change code to only remove those rows with NAs in necessary columns, prior to modeling the data. This is related to Issue #19.
- [x] Change some identifier names of inner-variables for the sake of comprehensibility and consistency. (But NOT changing any of the variables that are e.g. arguments of or returned by the model functions, because these names have already been released to the public and documented as canon.) Example of some changes are : the identifiers `dataList`, `POI`, and `genInitList` (which are created only to be passed to Stan, and thus are invisible to the plain user) have been renamed to be `stan_data`, `stan_pars`, and `stan_init`.
- [x] Move the message box "Extract model-based regressors" to a lower part of the code, so that all `stop()` checks are completed first before any of the actual processes begin or process messages are printed.
- [ ] If user-given data does not have all the correct data columns, print a message for the user what are the necessary data_columns for using this model, for the convenience of the user.
- [ ] Rearrange and combine the different parts of code for "", "single", "multipleB" type models for efficiency and comprehensibility
- [ ] A new general function to create `stan_init` (previously `genInitList`)

### Suggestions
1. The current hBayesDM `modelData` object returned as a result of running the code indexes the subjects of the user-data by the order of their appearance. So if a user were to input data with subjects in the order of 4, 3, 2, 1, then the returned `modelData` would yield data labeled as "1, 2, 3, 4" but here "1" would indicate subject 4, "2" would indicate subject 3, "3" would indicate subject 2, and so on. The suggestion is to sort() the subject names(whether labeled as numbers like "1, 2, 3, 4, ..." or names like "Alex, Beth, Cindy, ...") and then assign indexes to them in ascending order. Bench-marking tests of my own, and from [here] (https://stackoverflow.com/questions/17223308/fastest-way-to-count-occurrences-of-each-unique-element/17223496) show that changing the R code to use `rle(sort(a_vector))` is quicker than current implementations.
2. Add a process to prompt the user to ask [y/n] if user wants to omit NA rows and continue to the fitting, or stop and re-check their data (in the case NA rows are found.) Current implementation is to always omit NA rows and continue (with a warning message.)
3. Change to print the list of actual line numbers containing NA and thus omitted e.g. `lines : 12, 39, 112`, instead of (like the current implementation to) only print the amount of lines omitted e.g. `3 lines`.
4. Make the data column labels of the user-given data case-insensitive, so that if user's data has "subjid" instead of "subjID" it still works.

**Comment by JaeyeongYang** [comment]
@dlemfh96 
1. Great.
2. I'm not sure. Using a prompt makes it not a pure function. I think it'd be better to use a deterministic policy (e.g., to omit NA rows with warnings, or to stop the function with an error message).
3. Great. But, when this one is combined with the first suggestion, I'm afraid that specific line numbers do not have much meaning...
4. Great. How about using snakecase column names (e.g., `subj_id`), too?

**Comment by dlemfh96** [comment]
#### (See above Suggestions for more details)
- [ ] Sort subjects before passing to Stan
- [ ] ~~Prompt user before removing NA rows~~
- [ ] Print the actual line numbers omitted because of NA. (This is done in the code before the sorting of subjects, so this information will still be relevant to the user!)
- [x] Make necessary data column names *case-insensitive* and also *insensitive to underscores*. So that `"subjID"`, `"SubjID"`, `"subj_id"` all work.

**Comment by dlemfh96** [comment]
## Counting the number of trials for each subject
- The current implementation makes a list of unique subjects, then for each subject, searches the whole data set for rows belonging to that subject. Thus if the # of subjects is N, and # of rows of the whole data set is M, this current implementation would take O(N * M) searches.
- The suggested way to do this is to use `rle(sort(a_vector))`. This means to first sort the "subjID" column vector of whole data, which would take O(M * log(M)). Then, because it is sorted, getting the [run-length-encoding](https://en.wikipedia.org/wiki/Run-length_encoding) of that sorted vector would give us the number of trials for each subject. Since `rle()` only takes O(M), the total would be O(M * log(M)).

Using these terms, we could say that up to somewhere around log(M)<=N, the code that sorts is actually faster than the current implementation of not sorting. The example data set that I ran bench-marking tests on is "ra_data_attend.txt", which has 30(=N) subjects and 4000(=M) rows. Which means log(M) is 3.x, which is much smaller than N(=30). And thus the bench-marking tests also had the same result, of sorting being faster than not sorting.

- Another way to get the # of trials for each subject is to use the library "data.table". Which is the fastest out of all seemingly possible ways to do this, according to tests done [here](https://stackoverflow.com/questions/17223308/fastest-way-to-count-occurrences-of-each-unique-element/17223496). But the hBayesDM package would have an added dependency for the "data.table" library, and the difference in results may not be anywhere near drastic.

#### Thus I would need your opinions and consent before choosing which implementation to go with! @JaeyeongYang  @youngahn

**Comment by youngahn** [comment]
It’s okay to have dependency for the package, so maybe use the data.table?

**Comment by JaeyeongYang** [comment]
Solved with #45.
